src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c

```diff
@@ -928,19 +896,335 @@ int xocl_cleanup_mem(struct xocl_drm *drm_p)
 	return ret;
 }
 
+static int xocl_cleanup_drm_memory_manager(struct xocl_mm *xocl_mm)
+{
+	int i = 0;
+
+	for (i = 0; i < xocl_mm->m_count; i++) {
+		if (xocl_mm->mm_usage_stat && xocl_mm->mm_usage_stat[i])
+			vfree(xocl_mm->mm_usage_stat[i]);
+	}
+
+	vfree(xocl_mm->mm_usage_stat);
+
+	if (xocl_mm->bo_usage_stat)
+		vfree(xocl_mm->bo_usage_stat);
+
+	vfree(xocl_mm->mm);
+	vfree(xocl_mm);
+
+	return 0;
+}
+
+static int xocl_init_drm_mm(struct xocl_drm *drm_p, uint64_t mm_start_addr,
+		uint64_t mm_end_addr, uint32_t m_count)
+{
+	struct xocl_mm *xocl_mm = NULL;
+	size_t size = 0;
+	int err = 0;
+	int i = 0;
+	
+	BUG_ON(!mutex_is_locked(&drm_p->mm_lock));
+
+	xocl_mm = vzalloc(sizeof(struct xocl_mm));
+	if (!xocl_mm) {
+		return -ENOMEM;
+	}
+
+	size = m_count * sizeof(void *);
+	xocl_mm->mm_usage_stat = vzalloc(size);
+	if (!xocl_mm->mm_usage_stat) {
+		err = -ENOMEM;
+		goto error;
+	}
+
+	/* Initialize memory status for the memory manager */
+	xocl_mm->bo_usage_stat = vzalloc(XOCL_BO_USAGE_TOTAL *
+			sizeof(struct drm_xocl_mm_stat));
+	if (!xocl_mm->bo_usage_stat) {
+		err = -ENOMEM;
+		goto error;
+	}
+
+	for (i = 0; i < m_count; i++) {
+		xocl_mm->mm_usage_stat[i] = vzalloc(sizeof(struct drm_xocl_mm_stat));
+		if (!xocl_mm->mm_usage_stat[i]) {
+			err = -ENOMEM;
+			goto error;
+		}
+	}
+
+	/* Initialize the memory manager */
+	xocl_mm->mm = vzalloc(sizeof(struct drm_mm));
+	if (!xocl_mm->mm) {
+		err = -ENOMEM;
+		goto error;
+	}
+
+	drm_mm_init(xocl_mm->mm, mm_start_addr, (mm_end_addr - mm_start_addr));
+	if (!xocl_mm->mm) {
+		xocl_err(drm_p->ddev->dev,
+				"init memory manager failed");
+		err = -EINVAL;
+		goto error;
+	}
+
+	xocl_mm->start_addr = mm_start_addr;
+	xocl_mm->end_addr = mm_end_addr;
+	xocl_mm->m_count = m_count;
+
+	xocl_info(drm_p->ddev->dev, "drm_mm_init called for the available memory range"
+			" <%llx - %llx>", mm_start_addr, mm_end_addr);
+
+	drm_p->xocl_mm = xocl_mm;
+
+	return 0;
+
+error:
+	xocl_cleanup_drm_memory_manager(xocl_mm);
+	return err;
+}
+
+static int xocl_init_drm_memory_manager(struct xocl_drm *drm_p)
+{
+        struct xocl_dev *xdev = drm_p->xdev;
+	struct xocl_mm *xocl_mm = NULL;
+        void *blob = NULL;
+        int offset = 0;
+        const u64 *prop;
+        uint64_t mem_start, mem_end;
+        const char *ipname;
+        uint64_t mm_start_addr = 0;
+        uint64_t mm_end_addr = 0;
+        int err = 0;
+
+        blob = XDEV(xdev)->fdt_blob;
+        if (!blob)
+                return 0;
+
+	mutex_lock(&drm_p->mm_lock);
+	xocl_mm = vzalloc(sizeof(struct xocl_mm));
```

Comment:
memory leak? I don't see xocl_mm being assigned to drm_p->xocl_mm?

Reply:
You are correct. Fixed this issue.